### ISSUE TITLE
false positive: prefer_constructors_over_static_methods should not report on method which can return null

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#2280]
**Describe the issue**

Under Null safety, factory constructors can no longer return null. So prefer_constructors_over_static_methods should not suggest that a static method which can return null should be converted to a constructor.

**Examples**

```dart

class A {
  static A x() => null;

  static A y() {
    if (1==1) return A();
    else return null;
  }

  static A? z() {
    return aOrNull;
  }
}

A? get aOrNull => 1 == 2 ? A() : null;

```

**To Reproduce**

The examples above

**Expected behavior**

No lint reported on any of the methods.

**Additional context**

I think the implementation would be split in two:

* In a pre-null safety library, this lint should check for literal null returns
* In a null safety library, we could just skip methods which return a nullable type.
